### PR TITLE
New Invite Screen

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
Added new invites screen and invite URL support.

cc: @vvisigoth @jyng

### CODE REVIEW

File: src/form/EmailChipInput.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  );
+
+  const addToChips = useCallback(() => {
+    setChips([...chips, value]);
+    setValue('');
+  }, [chips, value, setChips]);
+
+  const handleKeyDown = useCallback(
+    event => {
+      if (['Enter', 'Tab', ','].includes(event.key) && value !== '') {
```
Comment (Fang-) [inline src/form/EmailChipInput.js]: Can we make space do this too? ime that's commonly supported, and it doesn't have conflicts with standard/expected behaviors like tab and return do (next field and submit respectively).

File: src/components/InviteSigilList.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      Math.max(6 - _acceptedPoints.length - _pendingPoints.length, 0)
+    ).keys(),
+  ].map(() => Nothing());
+
+  const renderSigil = (points, colors, klassName) => {
+    return (
+      <>
+        {points.map((point, idx) => (
+          <div key={idx} className={cn(klassName, 'h9 w9')}>
+            <MaybeSigil patp={point} size={50} colors={colors} />
```
Comment (Fang-) [inline src/components/InviteSigilList.js]: As per design feedback, we should at least `title={point}` here.

File: src/views/Invite/InviteEmail.js [hunk 3]
```diff
@@ -142,15 +142,12 @@ export default function InviteEmail() {
...
-      }
-    },
-    [getHasReceived]
-  );
+  const validateHasReceived = useCallback(async email => {
+    // const hasReceived = await getHasReceived(email);
+    // if (hasReceived) {
+    // return HAS_RECEIVED_TEXT;
+    // }
+  }, []);
```
Comment (Fang-) [inline src/views/Invite/InviteEmail.js]: Also what's this about? This same logic now lives in `Views/Invite/Email.js` it seems, so maybe a remnant of the move?

File: src/components/InviteSigilList.js (hunk 2 above)
Comment (liam-fitzgerald) [reply src/components/InviteSigilList.js]: done

File: src/form/EmailChipInput.js (hunk 1 above)
Comment (liam-fitzgerald) [reply src/form/EmailChipInput.js]: done

File: src/views/Invite/InviteEmail.js (hunk 3 above)
Comment (liam-fitzgerald) [reply src/views/Invite/InviteEmail.js]: reverted these changes, but maybe better to just turf the old views

Comment (Fang-) [reply src/views/Invite/InviteEmail.js]: If they're unused they should just go.

File: src/lib/useBlockWindowClose.js [hunk 4]
```diff
@@ -15,3 +15,19 @@ export default function useBlockWindowClose() {
     };
   });
 }
+
+export function useConditionalBlockWindowClose(condition) {
```
Comment (Fang-) [inline src/lib/useBlockWindowClose.js]: This doesn't keep me from going to a different page within Bridge though. Is that something we can prevent?

**Comment by Fang-** [comment]
Changes seem good. Waiting on @jyng for design things.

**Comment by jyng** [comment]
> On the Point page, the sigils feel pretty visually heavy compared to the rest of the page. I'm not sure what we get out of them being there right now. At the very least, maybe they should show names on-hover (with just a title attribute)?

This makes the interface quite cluttered and the added benefit is quite trivial. The designs here are also incorrect — the row shouldn't exist inside the invite group view.

> Also the "pending" highlighting is the only color we have on the page, but it's not really the most important part of the page. Perhaps we want to use grey instead of orange?

Adding members is an important feature we want to highlight, so when people add members they should see positive feedback. Whether or not the color works is a different conversation.

> On the Invite Group screen, the sigils are redundant with the sigil+name list that's underneath them. We probably shouldn't be rendering them there. It's kind of weird, too, that I can't send invites from that dedicated screen, even though I can from the smaller overview on the Points page.

 The sigil list is the detailed version of the sigil row from the previous screen. There’s inaccuracy here: the row should only exist in the home screen not the invite group screen. And yes ‘Add Members’ should be in the Invite Group Screen.

> On the Point page, we might want a separator underneath the "add members" section. It feels a little too intimate with the "admin" button (and the rest of that section) right now.

Yeah this is a good point. We should probably drop it two 2 steps in our spacing scale

**Comment by jyng** [comment]
@liam-fitzgerald 
Some design fixes to be made:
Home page
- 'View' on hover should be a cursor
- Add Members should be a center button

Invite Group page
- There shouldn't be a sigil row
- There should be an Add Members button

Attached designs for reference

home sent
(Ignore the simplified navigation at the bottom, thats still work in progress)

invitegroup

**Comment by Fang-** [comment]
> This makes the interface quite cluttered and the added benefit is quite trivial. The designs here are also incorrect — the row shouldn't exist inside the invite group view.

Fair, but that doesn't answer my concerns about this row existing in the first place.

> Adding members is an important feature we want to highlight, so when people add members they should see positive feedback.

Hmm I'm not sure I understand your translation from "adding members" to "pending invites" here. I get that they're causally related, and I agree that the action of adding members is worth highlighting (as it already is, there's a big black button on an otherwise mostly white page), but I don't see "pending invites" as equally important.

**Comment by jyng** [comment]
Hmm, your questions seem to be regarding why 'invite group' is so emphasized. Let's move this into an email thread, your concern about the design is separate from Liam's work.

File: src/lib/useBlockWindowClose.js (hunk 4 above)
Comment (liam-fitzgerald) [reply src/lib/useBlockWindowClose.js]: I'd need to add lock/unlock functionality to the router

**Comment by jyng** [comment]
@liam-fitzgerald 
I'm noticing in this PR that the tab spacing is off. in Tabs.js line 36 i see we're setting margin/padding, but it's not clear to me why this is for? The spacing seems fine without it.

**Comment by liam-fitzgerald** [comment]
> @liam-fitzgerald
> I'm noticing in this PR that the tab spacing is off. in Tabs.js line 36 i see we're setting margin/padding, but it's not clear to me why this is for? The spacing seems fine without it.

Can you clarify this? I haven't changed the margin/padding in Tabs.js

**Comment by jyng** [comment]
Yeah you haven't changed any spacing on this PR, but I had just noticed this and was curious what it's for because it creates an inset on the tab bar.

image

**Comment by jyng** [comment]
Design Notes on Invite Screen
- The invite number should be 0 for pending invites
- You should see the pending invites when you click into Invite Group
- The textarea should be width: 100%
- We shouldn’t have the hover labels?? It’s just quite clunky

**Comment by Fang-** [comment]
> We shouldn’t have the hover labels?? It’s just quite clunky

They're pretty big, I was thinking just native `title` tag hovers. But then, if we don't have those, what do we even want to accomplish with having the sigils there?

**Comment by liam-fitzgerald** [comment]
> Yeah you haven't changed any spacing on this PR, but I had just noticed this and was curious what it's for because it creates an inset on the tab bar.
> 
> image

It was to center the tab options to match the designs. I'll get rid of that inset

**Comment by jyng** [comment]
> They're pretty big, I was thinking just native title tag hovers. **But then, if we don't have those, what do we even want to accomplish with having the sigils there?**

Just getting a sense of activity inside your invite group is enough information here, the name is optional. However, I think using the title attribute is an improvement here, so let's keep it.

**Comment by jyng** [comment]
LGTM

File: src/views/Invite/Cohort.js [hunk 5]
```diff
@@ -181,7 +181,7 @@ export default function InviteCohort() {
         <Grid.Item full>
           <Flex align="center">
             <Flex.Item>
-              {_acceptedInvites + _pendingInvites} / {_totalInvites}
```
Comment (Fang-) [inline src/views/Invite/Cohort.js]: Huh, I thought this was correct, in that these are the amount of invites you've actually _used_. idk though.

File: src/views/Invite/Url.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            progress={0.5}>
+            Generating Invite URL
+          </Grid.Item>
+        );
+      default:
+        return null;
+    }
+  }, [txStatus, error, generateInvite]);
+
+  const url = `https://bridge.urbit.org/#${get(invites, '[0].ticket', '')}`;
```
Comment (Fang-) [inline src/views/Invite/Url.js]: This includes the leading `~`. We've traditionally (with the emails) been sending out invites in the format of `bridge.urbit.org/#some-ticket`, without leading `~`. I think it looks cleaner, and would remove it for the sake of consistency, but don't feel super strongly.

Comment (vvisigoth) [reply src/views/Invite/Url.js]: Agree with @Fang- here, but that's my only quibble.

**Comment by vvisigoth** [comment]
I bless.

**Comment by vvisigoth** [comment]
I bless.

File: src/views/Invite/Url.js (hunk 6 above)
Comment (liam-fitzgerald) [reply src/views/Invite/Url.js]: fixed

**Comment by liam-fitzgerald** [comment]
ping

**Comment by Fang-** [comment]
Merging blessed PR!

**Pull request merged**